apis/v1alpha2/validation/gateway_test.go

```diff
@@ -141,6 +143,57 @@ func TestValidateGateway(t *testing.T) {
 			},
 			expectErrsOnFields: []string{"spec.listeners[0].hostname"},
 		},
+		"tls config present with udp protocol": {
+			mutate: func(gw *gatewayv1a2.Gateway) {
+				gw.Spec.Listeners[0].Protocol = gatewayv1a2.UDPProtocolType
+				gw.Spec.Listeners[0].TLS = &tlsConfig
+			},
+			expectErrsOnFields: []string{"spec.listeners[0].tls"},
+		},
+		"tls config present with http protocol": {
+			mutate: func(gw *gatewayv1a2.Gateway) {
+				gw.Spec.Listeners[0].Protocol = gatewayv1a2.HTTPProtocolType
+				gw.Spec.Listeners[0].TLS = &tlsConfig
+			},
+			expectErrsOnFields: []string{"spec.listeners[0].tls"},
+		},
+		"tls config present with tcp protocol": {
+			mutate: func(gw *gatewayv1a2.Gateway) {
+				gw.Spec.Listeners[0].Protocol = gatewayv1a2.TCPProtocolType
+				gw.Spec.Listeners[0].TLS = &tlsConfig
+			},
+			expectErrsOnFields: []string{"spec.listeners[0].tls"},
+		},
+		"hostname present with tcp protocol": {
+			mutate: func(gw *gatewayv1a2.Gateway) {
+				hostname := gatewayv1a2.Hostname("foo.bar.com")
+				gw.Spec.Listeners[0].Hostname = &hostname
+				gw.Spec.Listeners[0].Protocol = gatewayv1a2.TCPProtocolType
+			},
+			expectErrsOnFields: []string{"spec.listeners[0].hostname"},
+		},
+		"hostname present with udp protocol": {
+			mutate: func(gw *gatewayv1a2.Gateway) {
+				hostname := gatewayv1a2.Hostname("foo.bar.com")
+				gw.Spec.Listeners[0].Hostname = &hostname
+				gw.Spec.Listeners[0].Protocol = gatewayv1a2.UDPProtocolType
+			},
+			expectErrsOnFields: []string{"spec.listeners[0].hostname"},
+		},
+		"wildcard hostname": {
+			mutate: func(gw *gatewayv1a2.Gateway) {
+				hostname := gatewayv1a2.Hostname("*")
+				gw.Spec.Listeners[0].Hostname = &hostname
+			},
+			expectErrsOnFields: []string{"spec.listeners[0].hostname"},
+		},
+		"empty string hostname": {
+			mutate: func(gw *gatewayv1a2.Gateway) {
+				hostname := gatewayv1a2.Hostname("")
+				gw.Spec.Listeners[0].Hostname = &hostname
+			},
+			expectErrsOnFields: []string{"spec.listeners[0].hostname"},
+		},
```

Comment:
Now that we're using regex validation on the hostname, we likely don't need the validation in the webhook. With that said, it is much nicer to be able to use upstream validation functions included here. Since this validation wasn't actually added in this PR, I don't think this needs to block anything, I've added https://github.com/kubernetes-sigs/gateway-api/issues/868 to track this.